purchases/src/main/kotlin/com/revenuecat/purchases/coroutines.kt

```diff
@@ -0,0 +1,26 @@
+package com.revenuecat.purchases
+
+import com.revenuecat.purchases.interfaces.ReceiveCustomerInfoCallback
+import kotlin.coroutines.resume
+import kotlin.coroutines.suspendCoroutine
+
+/**
+ * Get latest available purchaser info.
+ * Coroutine friendly version of [Purchases.getCustomerInfo].
+ *
+ * @return [Result] type containing either the [CustomerInfo] or the [PurchasesError]
+ */
+suspend fun Purchases.getCustomerInfoSuspend(): Result<CustomerInfo, PurchasesError> {
+    return suspendCoroutine { continuation ->
+        val receiveCustomerInfoCallback = object : ReceiveCustomerInfoCallback {
+            override fun onReceived(customerInfo: CustomerInfo) {
+                continuation.resume(Result.Success(customerInfo))
```

Comment:
I am not convinced about the fact that this is using our own `Result`. There's a `Result` class in Kotlin already and the IDE gets super confused.

We could change it to just be `resume(customerInfo)`, and I think ideally we would call `resumeWithException`, but I am not sure how to convert the `PurchasesError` to an `Exception`. 

Any ideas @tonidero @pablo-guardiola

Reply:
Yeah, I think with coroutines, the return value should be the "success" value and any errors should be handled as exceptions. I do have doubts in some places where errors can be common, for example, when a user cancels a purchase, I guess that should be considered an exception, but it will be pretty common.

As for how to convert a `PurchasesError` to an Exception, I'm thinking of 3 ways:
- Create a custom exception that has a `PurchasesError` property. This allows us to keep our API for coroutines and not callbacks pretty similar, but our public API would have an extra indirection level, like `exception.purchasesError.code`
- Copy the properties of `PurchasesError` to the new exception. The API for coroutines will be more straightforward but will differ somewhat from the callbacks API.
- A mix of both, have the exception contain the `PurchasesError` but also provide convenience methods to access the properties directly.

I guess I'm kinda leaning towards the third option, thoughts?

Reply:
I agree with @tonidero it's preferable to optimize for the happy path and return errors as exceptions otherwise - coroutines exception handling is 💪

I also think that the third option is the way to go as it's the best of both worlds and it'll provide an easy to use API. I'd remove the suffix `Suspend` from the name though and will  rename it to `awaitCustomerInfo`, wdyt?

Another thing to consider when adding coroutines support is to have non-suspendable APIs that return a `Flow`.